apollo-router/src/plugins/authentication/subgraph.rs

```diff
@@ -192,12 +198,100 @@ pub(crate) struct SigningParams {
     region: Region,
     service_name: String,
     subgraph_name: String,
 }
 
+#[derive(Clone, Debug)]
+struct CredentialsProvider {
+    credentials: Arc<RwLock<Credentials>>,
+    _credentials_updater_handle: Arc<JoinHandle<()>>,
+    #[allow(dead_code)]
+    refresh_credentials: Sender<()>,
+}
+
+// Refresh token if it will expire within the next 5 minutes
+const MIN_REMAINING_DURATION: Duration = std::time::Duration::from_secs(60 * 5);
+
+impl CredentialsProvider {
+    async fn from_provide_credentials(
+        provide_credentials: impl ProvideCredentials + 'static,
+    ) -> Result<Self, CredentialsError> {
+        let credentials_provider = SharedCredentialsProvider::new(provide_credentials);
+        let (sender, mut refresh_credentials_receiver) = tokio::sync::mpsc::channel(1);
+        let credentials = credentials_provider.provide_credentials().await?;
+        let mut refresh_timer = next_refresh_timer(&credentials);
+        let credentials = Arc::new(RwLock::new(
+            credentials_provider.provide_credentials().await?,
+        ));
+        let c2 = credentials.clone();
+        let handle = tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = tokio::time::sleep(refresh_timer) => {
+                        // Refresh credentials
+                        if let Ok(new_credentials) = credentials_provider.provide_credentials().await {
+                            refresh_timer = next_refresh_timer(&new_credentials);
+                            let mut credentials = c2.write().unwrap(); // todo: unwrap
+                            *credentials = new_credentials;
+                        }
+                    },
+                    rcr = refresh_credentials_receiver.recv() => {
+                        if rcr.is_some() {
+                            if let Ok(new_credentials) = credentials_provider.provide_credentials().await {
+                                let mut credentials = c2.write().unwrap(); // todo: unwrap
+                                *credentials = new_credentials;
+                            }
+                        } else {
+                            return;
+                        }
+                    },
+                }
+            }
+        });
+        Ok(Self {
+            _credentials_updater_handle: Arc::new(handle),
+            refresh_credentials: sender,
+            credentials,
+        })
+    }
+
+    #[allow(dead_code)]
+    pub(crate) async fn refresh_credentials(&self) {
+        let _ = self.refresh_credentials.send(()).await;
+    }
+}
+
+fn next_refresh_timer(credentials: &Credentials) -> Duration {
+    credentials
+        .expiry()
+        .and_then(|expiry| {
+            expiry
+                .duration_since(SystemTime::now())
+                .ok()
+                .and_then(|d| d.checked_sub(MIN_REMAINING_DURATION))
+        })
+        .unwrap_or_else(|| Duration::from_secs(0))
```

Comment:
This will cause the router to try to refresh credentials in a loop if there are no expiry value. Long-lived credentials (e.g. AWS IAM Users credentials) won't have an expiry value, which would cause this to spin in a loop.

Should we treat credentials that never expire differently?

Reply:
Great question! We could use the default (15 minutes), which shouldn't put too much pressure on the aws systems.